### PR TITLE
fix: Generate shrinkwraps for the bundled vscode

### LINKED ISSUE #5056: [Bug]: Use npm instead of yarn in postinstall script
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS/Web Information

- Web Browser: n/a
- Local OS: n/a
- Remote OS: n/a
- Remote Architecture: n/a
- `code-server --version`: development

### Steps to Reproduce

1. `yarn build release`
2. Uninstall yarn
3. `cd release && npm install`

### Expected

Install completes successfully.

### Actual

Although I have not actually done this (uninstalling yarn is a pain), the postinstall script uses yarn so it will probably error that yarn cannot be found.

### Logs

_No response_

### Screenshot/Video

_No response_

### Does this issue happen in VS Code?

- [X] I cannot reproduce this in VS Code.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

### Notes

To fix:

1. Replace `yarn` with `npm` in `npm-postinstall.sh`
2. Generate shrinkwrap files for:
    1. `lib/vscode/package.json`
    2. `lib/vscode/extensions/package.json`

### PR DESCRIPTION
**Pull request by edvincent**
Fixes #5056 #5174

> **Why are we removing `synp` after all?**

Because it only takes care of putting in the package-lock/shrinkwrap file the first level of dependencies, not the nested dependencies which we have from the yarn lockfile.

See the contents of https://unpkg.com/browse/code-server@4.2.0-5047-90c9adcded65d10004a9db71d0717cd90565baa3/npm-shrinkwrap.json, and how for example, `@mapbox/node-pre-gyp` is listed as a dependency of `argon2` - but is not present anywhere in the shrinkwrap file.

> **Why are we adding packages from `resolutions` in `devDependencies` too?**

What initially drove me to use `synp` was the fact that the `npm shrinkwrap` command would generate a lockfile that wasn't usable. Trying to install the package with that lockfile would fail with the following exception:

```
Installing Code dependencies...
[..]
npm ERR! Cannot read property 'requires' of undefined
```

This was actually due to `npm shrinkwrap` (in NPM 6) choking with some packages not being anywhere in the lockfile but being present in the `package.json`.

So I simply tracked down their dependency tree, saw they were coming from devDependencies, and added them there - which now prevents the `npm shrinkwrap`-generated lockfile from breaking, and allows us to remove `synp`. This doesn't change the size of the dependency tree, and should be able to be removed when there are higher versions of NPM used.

> **Why aren't we using `npm ci` to replace `yarn --frozen-lockfile`?**

NPM 6's `npm ci` doesn't deal very well with optional dependencies, which the bundled vscode needs for some Windows dependencies (namely, `windows-process-tree` and `@vscode/windows-registry`):
* The `npm shrinkwrap` command doesn't add them to the lockfile, because the CI runs from Linux, so it doesn't install them - and the commands generates the lockfile based on what's in `node_modules`.
* If we hacked the lockfile and added them there manually (with `"optional": true`), the install on a Linux machine would fail - because NPM 6's `npm ci` doesn't respect them from an `npm-shrinkwrap.json` file.
* If we don't add them and use `npm ci`, they are not installed - even on Windows machines.

What I settled for was using `npm install --production` (for now), that will always respect the lockfile version if an entry is there, and will try to get versions not listed there. This effectively means that **those two packages, on windows, won't be following the `yarn.lock` that is being used in the development package**, but that seemed less worse.

*Note: Another idea was to hackily add the packages to the lockfile + enhance the `postinstall.sh` script to conditionally use `--no-optional` if we're running on a non-Windows machine - which skips those optional dependencies. I wasn't convinced by either the hacking around of the lockfile nor this conditional... But happy to hear thoughts?*

> **Why don't we need the "Remove the `devDependencies`" hack anymore?**

Because we are now generating the shrinkwrap file after stuff has been installed with `yarn --production`, which means `npm shrinkwrap` will only list the production dependencies.

> **Why are we removing `json` after all?**
Because we don't need to remove the `devDependencies` anymore - which needed some more complex JSON mangling. Now we simply need to remove a line, which can be done with `sed`.

But also, because doing it with `json` gets a bit tricky, as `json` would need to be listed as an actual dependency (rather than `devDependency`) of code-server and vscode too to be able to be used... Not worth it.

Switching the `jq` commands to `json` might still be worth it, but no need to conflate the commits.

> **Why no more `yarn pack` artifact?**
Realized I was trying to fix/conflate too many things into one PR, so will likely create a different issue to discuss this one.

### CODE REVIEW

**Comment by edvincent** [comment]
So the failure seems to be because of some dependencies missing in the `node_modules` from the original install... Which my guess is because something is failing to be installed. Will be playing around with it (expect a bunch of updates to this PR to "test" the CI) and try to find the culprit.

**Comment by code-asher** [comment]
GitHub seems to have ignored my email response so I will paste it below, apologies if it does eventually come in and the comment gets duplicated:

----

I love the detailed description!

> What is used to install is completely meaningless to the end-user. If
> only, by not using yarn even when invoked with yarn, we can guarantee
> the deterministic versions for the bundled vscode. Thoughts?

My first thought when I mentioned this was that the user might have
`yarn` installed but not `npm` so switching the package manager from
under the user could cause installation failures (`npm: command not found` 
or something like that).

Buuuut...is it likely someone has Node installed but not npm?  I know
some distros package them separately but it still seems unlikely.

The second thought I had was that using the same package manager
throughout the entire installation process /feels/ more correct to me.
I doubt it would cause issues in practice but it seems like it would be
less of a surprise ("Wait why is it calling npm when I ran yarn?").

File: .github/workflows/ci.yaml
Comment (edvincent) [inline .github/workflows/ci.yaml:114]: I suspect this was the culprit of the failures: because `npm shrinkwrap` generates the shrinkwrap file from the existing `node_modules` folder (to extract the versions that had been installed), if an optional module failed to install (e.g. because some build dependencies like X11 etc.. were missing) - it wouldn't be in the shrinkwrap file.

Unfortunately, looks like I need some more approvals again to be able to run the workflows and validate this... @jsjoeio @code-asher if you can unblock the run(s) that'd be awesome!

File: ci/build/build-release.sh
Comment (code-asher) [inline ci/build/build-release.sh:168]: My bad; I should have made a note of this earlier; we recently switching to using `lib/vscode/remote/package.json` since it turns out the upper `package.json` is for desktop VS Code and one in `remote` is for the web version (which is what code-server is running).

Could we generate the shrinkwraps directly in `$RELEASE_PATH`?  That way it is always using the right `package.json` no matter what we end up copying into it.

Comment (edvincent) [reply ci/build/build-release.sh:168]: Yup let me try that. For now mostly trying to get it to build more than anything, so no probs on other things.

**Comment by edvincent** [comment]
Ok new tentative out, with:
1. Using `npm` or `yarn`, depending on what was initially used.
1. Generating the shrinkwraps when generating the release, rather than initially.
1. Removing `@parcel/node-addon-api` from the shrinkwrap, as it's a folder in `node_modules` which is actually not a package and was making the whole process choke.

(1) and (2) together (should) also help the other issue I was trying to track down, which happens because the tests were failing for the standalone release as it wasn't installing some of the devDependencies that we were removing too early from the shrinkwrap file.

**Comment by edvincent** [comment]
FWIW, I'm playing around in https://github.com/edvincent/code-server/pull/1#partial-pull-merging to be able to run the CI/CD and pinpoint some of the culprits. To avoid the flood and needing approvals for the workflows to run. Sorry it's taking so long, but trying to do it the best possible way this time 💪

**Comment by edvincent** [comment]
Ok, I think I got it this time! Moved to generate the shrinkwraps in the `release:standalone` command, which basically allows not having to know which `package.json` is being copied etc...

Also updated the overview of this PR to cover some of the reasoning.

Thoughts? I think I still need to add something so that the NPM publish CI step calls this release:standalone too?

**Comment by code-asher** [comment]
Awesome work!  I love your comments; very thorough and made it super easy to understand the rationale.

> I think I still need to add something so that the NPM publish CI step calls this release:standalone too?

I think leaving it in `build-release.sh` is the right move.  We use the result of that script in both the NPM publish step and standalone step.

**Comment by code-asher** [comment]
Ohh I see in your PR comment you are doing it there because it runs after `yarn --production`.  Hmm...

**Comment by edvincent** [comment]
Yep.. We need the `node_modules` to be populated to be able to generate the shrinkwrap file...

The other solution would be to still do it in `build-release.sh`, which means we'd need to `yarn install` in `build-release.sh`. That would address this `TODO` https://github.com/coder/code-server/blob/main/.github/workflows/ci.yaml#L240-L242 - and the NPM release would move to be a `yarn pack` (rather than a raw `tar` command). But I'm not sure how would having the `node_modules` folder from one arch would affect another arch - for the cross-compile sections...

Unless we do it in `build-release.sh`, with a `yarn install`, but then generate two type of artifacts - the current one (for the other platforms) and the current one...

Thoughts?

**Comment by code-asher** [comment]
Maybe we could just add `yarn --production` to `build-release.sh` and it would not take too much longer since it only needs to remove files (but I have no idea how yarn works internally so maybe this takes longer for some reason).

**Comment by code-asher** [comment]
Oh right we have no Node modules at all at that stage, we would need to remove that whole `KEEP_MODULES` behavior and just keep the modules (in addition to then running `yarn --production` at the end if I understand correctly).

**Comment by code-asher** [comment]
> But I'm not sure how would having the node_modules folder from one arch would affect another arch

This is exactly why I have avoided doing that todo so far lol

**Comment by edvincent** [comment]
Today there is never a `yarn` in `build-release.sh` - so a `yarn --production` won't be removing files, but also installing stuff that was never installed.

The only `yarn` that gets done is at the root of the repo, not inside the `release` folder no?

EDIT: Yep, what you just said now. Happy to try and take a stab at that if it seems to be the best.

**Comment by code-asher** [comment]
> Today there is never a yarn in build-release.sh

Ah yup it is done in a previous step and then this step has an option (`KEEP_MODULES`) for whether to also copy the Node modules or skip them.

**Comment by code-asher** [comment]
Yeah I say we go ahead with this plan.

**Comment by edvincent** [comment]
Damn it's actually a bit more nuanced than just removing the `KEEP_MODULES` behavior... If we run `yarn --production`, it would also install some of the things we were excluding in the `rsync` like the `node` executable, the `node_modules.ara` and the `lib/coder-cloud-agent`. Which means the `tar -czf` command in the `ci.yaml` would need to replicate/hard-code some of those excludes...

The other solution would be to come back to the original idea of generating the shrinkwrap files from the code-server repo itself, and copy them into release (after removing the devDependencies from it). i.e. we'd need to cd into `lib/vscode/remote/` - rather than relying on the fact we copied the right `package.json`.

BTW - Right now, that `KEEP_MODULES` never seems used?

**Comment by code-asher** [comment]
Er wait I might have misunderstood.  We do already run `yarn --production` so those things exist in the standalone already but now the npm package will include them when it did not before maybe?  In that case maybe we add to the npmignore.

**Comment by edvincent** [comment]
No, I haven't forgotten about this/you ❤️ life got in the way...

I'm back trying to fix this, sent a new approach PR (which gets a bit easier thanks to NodeJS 16!). Still a WIP and needs some tweaks, just sharing for awareness. Testing in https://github.com/edvincent/code-server/actions/runs/2883629793

HOWEVER, something that is a bit of a concern to me now seems to be a bug in NPM itself... Despite the shrinkwrap file, installing with the `global` flag still doesn't seem to respect them 😭

**Comment by edvincent** [comment]
FYI - opened https://github.com/npm/cli/issues/5325 to track that nasty bug about the shrinkwrap file not being respected.

I'll still continue with the PR here (need to clean-up some of the code/options, I used `--no-default-rc` everywhere but supposed to only be for one of the folders) and we likely will be able to merge it - as the result will be the same as the current state: every install installs the latest version. But as soon as the bug upstream will be fixed, stuff will start working with no action :p

Any thoughts on the approach proposed in the latest code?

File: ci/build/build-standalone-release.sh
Comment (code-asher) [inline ci/build/build-standalone-release.sh:36]: The `build-standalone-release.sh` script takes the npm package generated in a previous step, installs deps, then packages the result so we can upload a standalone build complete with deps to GitHub.  Assuming I have that right (sometimes I get confused with all the steps) generating a shrinkwrap here would not get uploaded to npm.

So I think we will have to do this in `build-release.sh` which is where we create and upload the npm package artifact that eventually gets uploaded to npm.

And then here we would install with `npm` instead of `yarn`.

File: ci/build/npm-postinstall.sh
Comment (code-asher) [inline ci/build/npm-postinstall.sh:128]: This is dope!

File: ci/build/npm-postinstall.sh
Comment (code-asher) [inline ci/build/npm-postinstall.sh:134]:
```suggestion
        echo "yarn.lock file not present, not running in development mode. use npm to install code-server!"
```

File: ci/build/build-standalone-release.sh
Comment (code-asher) [inline ci/build/build-standalone-release.sh:47]: > We don't need to remove the devDependencies for these

That makes sense, does that mean we should be removing them for the root dependencies?  It looks like we are including dev deps right now in the root shrinkwrap (I might be missing something!).

**Review by code-asher — COMMENTED** [review]
Welcome back from life!  :laughing:

This looks really good.

File: ci/build/build-standalone-release.sh
Comment (edvincent) [reply ci/build/build-standalone-release.sh:36]: Ah damn yes, now I think I remember being bugged/blocked by something like that in the past... Gonna double-check that now!

File: ci/build/build-standalone-release.sh
Comment (edvincent) [reply ci/build/build-standalone-release.sh:47]: So the expectation is that when installing something with `--global` infers only keeping the prod dependencies. So rather than manipulating the shrinkwrap to remove them, we should be clear to leave them there and let `npm` do its thing.

That being said, there seems to be another nasty bug with that... https://github.com/npm/cli/issues/4323

But I'm not sure that removing them from the shrinkwrap won't simply cause NPM to try to download them anyway... Let me play around!

Comment (code-asher) [reply ci/build/build-standalone-release.sh:47]: Ohhhh that makes sense!  Definitely agreed that we should just let `npm` do its thing.

File: ci/build/npm-postinstall.sh
Comment (edvincent) [reply ci/build/npm-postinstall.sh]: Ah yes good catch, thanks!

**Comment by edvincent** [comment]
Ok... This is what it looks like when doing it from the release script... Which is indeed where it's needed, as the tgz we use to publish are the artifacts from it - not from release standalone. Confirmed from the artifacts the shrinkwrap files and their content from downloading the build artifacts.

It looks a bit worse because shrinkwrapping tries to alter the `yarn.lock` - so we have to keep it and replace it again. But seems to be the least-worst way - the other being doing a `release:standalone` before publishing, which would need to install again all the development 
dependencies etc...

File: ci/build/build-standalone-release.sh
Comment (edvincent) [reply ci/build/build-standalone-release.sh]: > And then here we would install with `npm` instead of `yarn`.

Can I suggest to avoid doing too many things in the same PR if that works for you:
* (This PR) Just adding the `npm-shrinkwrap.json` file - which basically should not change anything else in the development/CI workflow.
* Another PR to change the `release:standalone` and other CI scripts to start using `npm ci`
* Another PR (when everything works) to amend the doc to recommend `npm` (which can be done last, as using yarn will still always work).

Comment (code-asher) [reply ci/build/build-standalone-release.sh]: I like the incremental approach!

File: ci/build/build-standalone-release.sh
Comment (code-asher) [inline ci/build/build-standalone-release.sh:2]: Looks like an extra hashbang snuck in.
```suggestion
```

**Review by code-asher — APPROVED** [review]
:tada:

File: ci/build/build-standalone-release.sh
Comment (edvincent) [reply ci/build/build-standalone-release.sh]: Yup - already sending an amend completely removing any diff on that file as not needed anymore.

Comment (code-asher) [reply ci/build/build-standalone-release.sh]: Awesome~

File: ci/build/npm-postinstall.sh
Comment (edvincent) [inline ci/build/npm-postinstall.sh:133]: This is just replicating the current behavior introduced in https://github.com/coder/code-server/pull/5135/commits/7579dbbb06ba6c4b9a3e9ad94cdccfde15fcb5c7

@jsjoeio Remember why it's needed for `lib/vscode` but not for `lib/vscode/extensions`? Happy to add a comment to explain why this is needed, likely useful :D

Comment (code-asher) [reply ci/build/npm-postinstall.sh:133]: My memory is fuzzy so @jsjoeio can confirm but I think it was because of one or more of the following:

- The `.yarnrc` (which targets Electron) is in `lib/vscode` so it might not affect `lib/vscode/extensions` (idk if yarn travels upwards looking for a `.yarnrc` but it might).
- Even if the `.yarnrc` does affect `lib/vscode/extensions`, that directory has no native modules so it would have no influence anyway

That said, there is probably no harm in always including `--no-default-rc` for all our `yarn` calls.

Comment (edvincent) [reply ci/build/npm-postinstall.sh]: > That said, there is probably no harm in always including `--no-default-rc` for all our `yarn` calls.

Reverting to that then! (Which will fix my bad `bash` syntax that just caused a failed build too).

Reasoning makes sense. And yes, `yarn` does travel the parent directories looking for other `.yarnrc` files - and merges them - as per https://classic.yarnpkg.com/en/docs/yarnrc/

Comment (code-asher) [reply ci/build/npm-postinstall.sh]: Aha that makes sense, thanks for the link!

Comment (jsjoeio) [reply ci/build/npm-postinstall.sh]: I should have added a comment because my memory is fuzzy too 😅 Sounds like @code-asher nailed it though.

**Comment by code-asher** [comment]
Everything looks good to me so I am going to merge it in.  Thanks a million!

**Comment by jsjoeio** [comment]
@edvincent really can't thank you enough - it's rare someone from the community takes on a hard problem, takes a break because life, then comes back to finish it 💪🏼 We appreciate you a ton!

**Comment by edvincent** [comment]
Anytime! I like giving back to the community too, especially for stuff I use daily! Plus I got to learn a LOT about NPM's dependency management, which I enjoyed. And thank to both of you for bearing with me during the (long) process.

You can also thank @bpmct and the nice t-shirt he sent me couple of months ago, it definitely played a role in reminding me I still had this PR pending every time I wore it. 👀

Will still send a follow-up PR (likely next week) for the documentation to remind folks to use `npm` rather than `yarn`.

**Comment by edvincent** [comment]
Also, whilst confirming the fix of https://github.com/coder/code-server/issues/5174#event-7234295102, realized that the bug https://github.com/npm/cli/issues/5325 is actually less of an issue here - because it doesn't seem to be buggy when installing it from a remote repo!

**Which means... There now is a way to install `code-server` with fully-deterministic dependencies, which will be the same whether `npm` or a packaged version is used. 🎉🎉**

**Comment by code-asher** [comment]
Awesome news!!

**Comment by code-asher** [comment]
I love the detailed description!

> What is used to install is completely meaningless to the end-user. If
> only, by not using yarn even when invoked with yarn, we can guarantee
> the deterministic versions for the bundled vscode. Thoughts?

My first thought when I mentioned this was that the user might have
`yarn` installed but not `npm` so switching the package manager from
under the user could cause installation failures (`npm: command not
found` or something like that).

Buuuut...is it likely someone has Node installed but not npm?  I know
some distros package them separately but it still seems unlikely.

The second thought I had was that using the same package manager
throughout the entire installation process /feels/ more correct to me.
I doubt it would cause issues in practice but it seems like it would be
less of a surprise ("Wait why is it calling npm when I ran yarn?").

**Comment by code-asher** [comment]
Ignore this lol GitHub seems to finally be posting comments I made via email months ago

**Pull request merged**
